mountpoint-s3/src/prefetch/parquet_prefetch.rs

```diff
@@ -0,0 +1,155 @@
+use crate::sync::Arc;
+use bytes::{Bytes, BytesMut};
+use futures::pin_mut;
+use futures::StreamExt;
+use intervaltree::IntervalTree;
+use mountpoint_s3_client::types::ETag;
+use mountpoint_s3_client::ObjectClient;
+use parquet::errors::ParquetError;
+use parquet::file::footer::decode_footer;
+use parquet::file::metadata::ParquetMetaData;
+use tracing::trace;
+
+use super::PrefetchReadError;
+
+const PARQUET_MAGIC_LEN: usize = 8;
+
+pub async fn read_parquet_metadata<Client>(
+    client: Arc<Client>,
+    bucket: &str,
+    key: &str,
+    if_match: &ETag,
+    total_size: u64,
+) -> Result<(Bytes, std::ops::Range<u64>), PrefetchReadError<Client::ClientError>>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    fetch_metadata(client, bucket, key, if_match, total_size)
+        .await
+        .map(|(metadata, footer, metadata_range)| {
+            let mut combined = BytesMut::with_capacity(metadata.len() + PARQUET_MAGIC_LEN);
+            combined.extend_from_slice(&metadata);
+            combined.extend_from_slice(&footer);
+            (combined.freeze(), metadata_range)
+        })
+        .map_err(|_| PrefetchReadError::MetadataParsingFailed)
+}
+
+async fn fetch_metadata<Client>(
+    client: Arc<Client>,
+    bucket: &str,
+    key: &str,
+    if_match: &ETag,
+    total_size: u64,
+) -> Result<(Bytes, [u8; PARQUET_MAGIC_LEN], std::ops::Range<u64>), ParquetError>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    if total_size < PARQUET_MAGIC_LEN as u64 {
+        return Err(ParquetError::General(
+            "Invalid Parquet file. Size is smaller than footer".to_string(),
+        ));
+    }
+
+    let mut footer = [0_u8; PARQUET_MAGIC_LEN];
```

Comment:
I see where some confusion is coming from. There's differences between what [Arrow say a Parquet footer is](https://arrow.apache.org/rust/parquet/file/footer/fn.decode_footer.html) and what [Parquet say a Parquet footer is](https://parquet.apache.org/docs/file-format/). 

I think we can solve this with a simple comment.

```suggestion
    // We refer to the last 8 bytes of a Parquet file as the footer.
    // This is sometimes described differently in different Parquet implementations.
    let mut footer = [0_u8; PARQUET_MAGIC_LEN];
```

Reply:
Yeah I agree, this is very confusing, and it was a mix up between the two sources like you mentioned. I'll add this comment to make it clearer!